tests/integration/test_scaling.py

```diff
@@ -0,0 +1,119 @@
+        mode = srvr(host)["Mode"]
+        assert mode in ["leader", "follower"]
+
+
+@pytest.mark.abort_on_fail
+async def test_deploy_active(ops_test: OpsTest):
+    charm = await ops_test.build_charm(".")
+    await ops_test.model.deploy(charm, application_name=APP_NAME, num_units=3)
+    await ops_test.model.block_until(lambda: len(ops_test.model.applications[APP_NAME].units) == 3)
+    await ops_test.model.set_config({"update-status-hook-interval": "10s"})
+    await ops_test.model.wait_for_idle(apps=[APP_NAME], status="active", timeout=1000)
+    for unit in range(3):
+        assert ops_test.model.applications[APP_NAME].units[unit].workload_status == "active"
+    await ops_test.model.set_config({"update-status-hook-interval": "60m"})
+
+
+@pytest.mark.abort_on_fail
+async def test_simple_scale_up(ops_test: OpsTest):
+    await ops_test.model.applications[APP_NAME].add_units(count=3)
+    await ops_test.model.block_until(lambda: len(ops_test.model.applications[APP_NAME].units) == 6)
+    await ops_test.model.wait_for_idle(apps=[APP_NAME], status="active", timeout=1000)
+    await ping_servers(ops_test)
+
+
+@pytest.mark.abort_on_fail
+async def test_simple_scale_down(ops_test: OpsTest):
+    await ops_test.model.applications[APP_NAME].destroy_units(
+        f"{APP_NAME}/5", f"{APP_NAME}/4", f"{APP_NAME}/3"
+    )
+    await ops_test.model.block_until(lambda: len(ops_test.model.applications[APP_NAME].units) == 3)
+    await ops_test.model.wait_for_idle(apps=[APP_NAME], status="active", timeout=1000)
+    await ping_servers(ops_test)
+
+
+@pytest.mark.abort_on_fail
+async def test_scale_up_replication(ops_test: OpsTest):
+    await ops_test.model.wait_for_idle(apps=[APP_NAME], status="active", timeout=1000)
+    await ping_servers(ops_test)
+    host = ops_test.model.applications[APP_NAME].units[0].public_address
+    model_full_name = ops_test.model_full_name
+    password = get_password(model_full_name or "")
+    write_key(host=host, password=password)
+    await ops_test.model.applications[APP_NAME].add_units(count=1)
+    await ops_test.model.block_until(lambda: len(ops_test.model.applications[APP_NAME].units) == 4)
+    await ops_test.model.wait_for_idle(apps=[APP_NAME], status="active", timeout=1000)
+    check_key(host=host, password=password)
+
+
+@pytest.mark.abort_on_fail
+async def test_kill_quorum_leader_remove(ops_test: OpsTest):
+    await ops_test.model.applications[APP_NAME].destroy_units(f"{APP_NAME}/0")
+    await ops_test.model.block_until(lambda: len(ops_test.model.applications[APP_NAME].units) == 3)
+    await ops_test.model.wait_for_idle(apps=[APP_NAME], status="active", timeout=1000)
+    await ping_servers(ops_test)
+
+
+@pytest.mark.abort_on_fail
+async def test_kill_juju_leader_remove(ops_test: OpsTest):
+    leader = None
+    for unit in ops_test.model.applications[APP_NAME].units:
+        if await unit.is_leader_from_status():
+            leader = unit.name
+            break
+
+    if leader:
```

Comment:
Annoyingly, this silly pattern was to avoid a type-hint warning, as 'techincally' any of the types before if leader: could be "" or None. In practicality, there'll always be a leader, it's just to reduce the annoying red errors.